jetty-core/jetty-websocket/jetty-websocket-core-client/src/main/java/org/eclipse/jetty/websocket/core/client/CoreClientUpgradeRequest.java

```diff
@@ -61,7 +61,7 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public abstract class CoreClientUpgradeRequest extends HttpRequest implements Response.CompleteListener, HttpUpgrader.Factory
+public abstract class CoreClientUpgradeRequest implements Response.CompleteListener, HttpUpgrader.Factory
```

Comment:
Shouldn't this implement `Request` or extend something like `Request.Wrapper`?

Reply:
@lachlan-roberts it was extending `HttpRequest`, but now `HttpRequest` is an internal class.
The only reason to extend was to also implement `HttpUpgrader.Factory`, and there was an `instanceof` check with special logic.
Now the `HttpUpgrader.Factory` is added as attribute to the request, so there is no strict need for this class to extend `Request` (which has a million methods to implement).

If you feel you want to give to this class the full `Request` API, then we can make it implement `Request`, but then we will have to reimplement all the methods.

I tried to export the internals of `jetty-client` to this module, but then I had horrible failures at OSGi level, which apparently it is not able to understand JPMS's `export ... to ...`, so the `Import-Package` directives were all lacking the necessary internal classes.
For this reason, I went for this current implementation, which feels better because it has a very minimal dependency on `jetty-client` classes (and works with OSGi).

Reply:
I noticed some of the methods have slightly different signature to the ones on `Request`, for example `timeout(long value, TimeUnit unit)` return value is different.

But if this class extends a `Request.Wrapper` (doesn't exist but is easy to generate), then it simplifies this class and stops it diverging from any further changes in the `Request` interface.

Reply:
I'll give it a go. There is the problem of the covariant return type, but I it's always been there.

Reply:
I also worry that we could miss some of the API from `Request` that people would like to use on this. 
Maybe we could just add a `getRequest()` method to access the `Request`?
But either way I think the _has-a_ relationship will work for now.